### PR TITLE
HBASE-23917 [SimpleRpcServer] Subsequent requests will have no response in case of request IO mess up

### PR DESCRIPTION
**Pull request by chenxu14**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 38s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   5m 16s |  master passed  |
| +1 :green_heart: |  compile  |   1m 25s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 40s |  master passed  |
| +1 :green_heart: |  shadedjars  |   4m 40s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  3s |  master passed  |
| +0 :ok: |  spotbugs  |   4m 31s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 36s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   5m 13s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 36s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   5m 23s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |  15m 58s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| +1 :green_heart: |  javadoc  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 54s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 56s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  |  72m  9s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   1m  5s |  The patch does not generate ASF License warnings.  |
|  |   | 138m 26s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.security.TestSecureIPC |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1231 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 82afb8159fea 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/Base-PreCommit-GitHub-PR_PR-1231/out/precommit/personality/provided.sh |
| git revision | master / 48a3ccf523 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/testReport/ |
| Max. process+thread count | 8874 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
Pardon me, I do not fully follow what is going on here. What do you mean by request IO mess up? Could you please explain more? How would it happen?

**Comment by chenxu14** [comment]
For example, client side sent a message like 4|ABCD, but server side decoding it to 4|ABC,
the exact reason is not clear yet, suspected that there are some bugs with the KeyValueCodec

File: hbase-server/src/test/java/org/apache/hadoop/hbase/ipc/TestBlockingIPC.java [hunk 1]
```diff
@@ -106,4 +121,62 @@ protected RpcServer createTestFailingRpcServer(Server server, String name,
...
+      BlockingInterface stub = newBlockingStub(client, rpcServer.getListenerAddress());
+      EchoRequestProto param = EchoRequestProto.newBuilder().setMessage(msg).build();
+      try {
+        stub.echo(null, param);
+        fail("RPC should have failed because request IO mess up.");
+      } catch (ServiceException e) {
+        assertTrue(e.toString(),
+            StringUtils.stringifyException(e).contains("RequestIOMessUpException"));
+        // mess up IO don't blocking subsequent requests
+        assertEquals(msg, stub.echo(null, param).getMessage());
```
Comment (wchevreuil) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/ipc/TestBlockingIPC.java]: nit: use _expected_ flag in _@Test_ annotation?

File: hbase-client/src/main/java/org/apache/hadoop/hbase/RequestIOMessUpException.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.io.IOException;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Thrown by a region server if the request IO mess up.
+ */
+@SuppressWarnings("serial")
+@InterfaceAudience.Public
+public class RequestIOMessUpException extends IOException {
```
[the remark this answers is not in the document]
Comment (chenxu14) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/RequestIOMessUpException.java]: rename it to RequestIOCorruptionException

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m  0s |  Docker mode activated.  |
| -1 :x: |  patch  |   0m  3s |  https://github.com/apache/hbase/pull/1231 does not apply to master. Rebase required? Wrong Branch? See https://yetus.apache.org/documentation/in-progress/precommit-patchnames for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/console |
| versions | git=2.17.1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 23s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 38s |  master passed  |
| +1 :green_heart: |  checkstyle  |   1m 30s |  master passed  |
| +1 :green_heart: |  spotbugs  |   2m 56s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 14s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 25s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   1m 27s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  11m 42s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1.  |
| +1 :green_heart: |  spotbugs  |   4m 35s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 25s |  The patch does not generate ASF License warnings.  |
|  |   |  40m 23s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1231 |
| Optional Tests | dupname asflicense spotbugs hadoopcheck hbaseanti checkstyle |
| uname | Linux 9a36f9386278 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 724f0478ed |
| Max. process+thread count | 94 (vs. ulimit of 12500) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) spotbugs=3.1.12 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  5s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   3m 59s |  master passed  |
| +1 :green_heart: |  compile  |   1m 22s |  master passed  |
| +1 :green_heart: |  shadedjars  |   6m  9s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  0s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   3m 48s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 23s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   6m  8s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 59s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 10s |  hbase-client in the patch passed.  |
| -1 :x: |  unit  | 212m 10s |  hbase-server in the patch failed.  |
|  |   | 241m 36s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.12 Server=19.03.12 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/artifact/yetus-jdk8-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/1231 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 595f21b91fae 4.15.0-91-generic #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 724f0478ed |
| Default Java | 1.8.0_232 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/artifact/yetus-jdk8-hadoop3-check/output/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/testReport/ |
| Max. process+thread count | 3738 (vs. ulimit of 12500) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-1231/1/console |
| versions | git=2.17.1 maven=(cecedd343002696d0abb50b32b541b8a6ba2883f) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by saintstack** [comment]
Closing. Most of patch is for code not used anymore (though a nice refactoring).

**Pull request closed without merging**